cli/crates/server/src/codegen_worker.rs

```diff
@@ -0,0 +1,93 @@
+use crate::{event::Event, types::ServerMessage};
+use common::environment::{Project, SchemaLocation};
+use std::{ffi, fs, path::Path, sync::OnceLock};
+use tokio::{
+    sync::{broadcast, mpsc::UnboundedSender},
+    task::JoinHandle,
+};
+
+static CODEGEN_WORKER_HANDLE: OnceLock<JoinHandle<()>> = OnceLock::new();
+
+/// Spawns the background worker responsible for generating TS type definitions for resolvers.
+pub(crate) fn start_codegen_worker(
+    server_events: broadcast::Receiver<Event>,
+    event_sender: UnboundedSender<ServerMessage>,
+) -> Result<(), ()> {
+    let handle = tokio::spawn(async move { codegen_worker_loop(server_events, event_sender).await });
+    CODEGEN_WORKER_HANDLE.set(handle).map_err(|_| ())
+}
+
+async fn codegen_worker_loop(
+    mut server_events: broadcast::Receiver<Event>,
+    event_sender: UnboundedSender<ServerMessage>,
+) {
+    let project = Project::get();
+    let resolvers_path = project.udfs_source_path(common_types::UdfKind::Resolver);
+    let schema_path = &project.schema_path;
+    let mut last_seen_sdl = None;
+    let generated_ts_resolver_types_path = project.grafbase_directory_path.join("generated/index.ts");
+
+    // Try generating types on start up.
+    if let SchemaLocation::Graphql(schema_path) = schema_path.location() {
+        if let Some(sdl) = read_sdl(schema_path, &mut last_seen_sdl) {
+            generate_ts_resolver_types(sdl, &generated_ts_resolver_types_path);
+        }
+    }
+
+    loop {
+        match server_events.recv().await {
+            Err(_err) => {
+                break; // channel is broken, we might as well stop trying
+            }
+            Ok(Event::Reload(path)) => {
+                let path = path.canonicalize().unwrap_or(path);
+                match schema_path.location() {
+                    SchemaLocation::Graphql(schema_path) if path == schema_path.as_path() => {
+                        // The SDL schema was edited. Regenerate types.
+
+                        let Some(sdl) = read_sdl(&path, &mut last_seen_sdl) else {
+                            continue;
+                        };
+
+                        generate_ts_resolver_types(sdl, &generated_ts_resolver_types_path);
+                    }
+                    _ if path.extension() == Some(ffi::OsStr::new("ts"))
+                        && path.ancestors().any(|ancestor| ancestor == resolvers_path) =>
+                    {
+                        // A resolver changed, check it.
+                        let Some(sdl) = last_seen_sdl.as_deref() else {
+                            continue;
+                        };
+                        if let Err(err) = typed_resolvers::check_resolver(sdl, &path) {
+                            event_sender
+                                .send(ServerMessage::CompilationError(format!("{err:?}")))
```

Comment:
Should there be a separate ServerMessage variant for this? The output looks good enough, I tested manually.